Src/Support/GoogleApis.Auth.PlatformServices_Shared/OAuth2/LocalServerCodeReceiver.cs

```diff
@@ -0,0 +1,393 @@
+                _listener.Start();
+                Port = ((IPEndPoint)_listener.LocalEndpoint).Port;
+            }
+
+            private TcpListener _listener;
+            private CancellationTokenSource _cts;
+
+            public int Port { get; }
+
+            public async Task<Dictionary<string, string>> GetQueryParamsAsync(CancellationToken cancellationToken = default(CancellationToken))
+            {
+                var ct = CancellationTokenSource.CreateLinkedTokenSource(_cts.Token, cancellationToken).Token;
+                using (TcpClient client = await _listener.AcceptTcpClientAsync())
+                {
+                    try
+                    {
+                        return await GetQueryParamsFromClientAsync(client, ct);
+                    }
+                    catch (ServerException e)
+                    {
+                        Logger.Warning("{0}", e.Message);
+                        throw e;
+                    }
+                }
+            }
+
+            private async Task<Dictionary<string, string>> GetQueryParamsFromClientAsync(TcpClient client, CancellationToken cancellationToken)
+            {
+                var stream = client.GetStream();
+
+                var buffer = new byte[1];
+                int bufferOfs = 0;
+                int bufferSize = 0;
+                Func<Task<byte?>> getByte = async () =>
+                {
+                    if (bufferOfs == bufferSize)
+                    {
+                        bufferSize = await stream.ReadAsync(buffer, 0, buffer.Length, cancellationToken);
+                        if (bufferSize == 0)
+                        {
+                            // End of stream
+                            return null;
+                        }
+                        bufferOfs = 0;
+                    }
+                    return buffer[bufferOfs++];
+                };
+
+                string requestLine = await ReadRequestLine(getByte);
+                var requestParams = ValidateAndGetRequestParams(requestLine);
+                await WaitForAllHeaders(getByte);
+                await WriteResponse(stream, cancellationToken);
+
+                return requestParams;
+            }
+
+            private async Task<string> ReadRequestLine(Func<Task<byte?>> getByte)
+            {
+                var requestLine = new StringBuilder(MaxRequestLineLength);
+                do
+                {
+                    if (requestLine.Length >= MaxRequestLineLength)
+                    {
+                        throw new ServerException($"Request line too long: > {MaxRequestLineLength} bytes.");
+                    }
+                    byte? b = await getByte();
+                    if (b == null)
+                    {
+                        throw new ServerException("Unexpected end of network stream reading request line.");
+                    }
+                    requestLine.Append((char)b.Value);
```

Comment:
(And if we expect everything to actually be ASCII, perhaps we should enforce that?)